### PR TITLE
Fix local bounds problems. Swap bounds with localBounds to do the trick.

### PR DESCRIPTION
**Pull request by SukantPal**
Fix for #6355 

By swapping the `_bounds` object with `_localBounds`, we are preventing `calculateBounds` from editing the world-bounds.

Also, I am syncing the changes in the `_boundsID` with the world-bounds `updateID`. This will prevent the world-bounds from getting recalculated in the next `getBounds(true)` if it was already update-to-date.

- [ ] Tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] Lint process passed (`npm run lint`)
- [ ] Tests passed (`npm run test`)

### CODE REVIEW

**Comment by SukantPal** [comment]
Hmm, this doesn't solve the JSFiddle I presented. On inspection, the `_bounds` of the object are correct now, except the sprite is still not rendering correctly.

**Comment by SukantPal** [comment]
The sprite still renders wrongly because `Container` will update the transform of its children too. The rendering process does not update transforms.

**Comment by SukantPal** [comment]
@ivanpopelyshev @bigtimebuddy This fixes the problem!

**Comment by SukantPal** [comment]
I think instead of reversing update-transform on the children in `Container#getLocalBounds`, I will change this so that we use `getLocalBounds` when calculating bounds of Container.

**Comment by SukantPal** [comment]
I think I will do one more optimization tomorrow: on a container's children, we should swap the transform with a temporary transform instance. Then we won't have to call update-transform again after `getLocalBounds`.

**Comment by bigtimebuddy** [comment]
It is important if you could capture this modified bounds behavior in a unit test, preferably one that breaks on dev. That’s the best way to ensure that we don’t have a regression.

File: packages/display/src/Container.ts [hunk 1]
```diff
@@ -479,6 +480,26 @@ export class Container extends DisplayObject
         this._bounds.updateID = this._boundsID;
     }
 
+    public getLocalBounds(rect?: Rectangle, noTransformSideEffects = true): Rectangle
```
Comment (eXponenta) [inline packages/display/src/Container.ts]: You override public method.
JSDoc is required.

`noTransformSideEffects` what this is mean and why this is `true` by default?

I think it is must be `skipUpdateTransfrom` how in getBounds method.

**Comment by ivanpopelyshev** [comment]
As I said in the issue:
1. I believe both of `getLocalBounds` in Container and DisplayObject should have `_` version of methods  because we have overrides in `Sprite` and there might be an override in spine library
2. I'm sure I can make a test case when this thing fails. I'll work on it in weekend
3. Try integrate my idea with axis-alighned fallback. However, with right architecture i can integrate it later.

File: packages/display/src/Container.ts (hunk 1 above)
Comment (SukantPal) [reply packages/display/src/Container.ts]: As I indicated before, this is just a sketch PR. That is why no JSDoc yet.

**Comment by SukantPal** [comment]
@ivanpopelyshev Your idea could be added in another PR - when it becomes your priority.

**Comment by SukantPal** [comment]
Overrides will not be affected at all. This should be a backward compatible change after I swap the child transforms too.

**Comment by SukantPal** [comment]
@ivanpopelyshev I’m looking for your tests today that should fail with this change.

**Comment by ivanpopelyshev** [comment]
OK, I wanted to merge dev into it but made something wrong. Can you please re-do it and push with force?

**Comment by ivanpopelyshev** [comment]
OK, never mind, I've rebased it myself.

**Comment by ivanpopelyshev** [comment]
We have to update all objects (Sprites, e.t.c.) `getLocalBounds` to pass that param. 

@bigtimebuddy we have a problem that `getBounds()` has first argument bool and second Rectangle, and for `getLocalBounds` its rectangle and then bool :) The heck we do in this case? I propose we add `skipChildrenUpdate` here as the first param. That way both methods will have the same signature and `false, null` by default.

File: packages/display/src/Container.ts (hunk 1 above)
Comment (SukantPal) [reply packages/display/src/Container.ts]: I removed that parameter.

**Comment by SukantPal** [comment]
@ivanpopelyshev Please recheck the standing of the changes.

**Comment by eXponenta** [comment]
> @ivanpopelyshev Please recheck the standing of the changes.

Doesn't work at all in real cases =)

DEV:
https://www.pixiplayground.com/#/edit/5uTdn6fP5w-DJrLXKuPWL

Your:
https://www.pixiplayground.com/#/edit/Qyl4KFWghAu3AwXfRLOSV

**Comment by SukantPal** [comment]
@eXponenta I already know. Review the types of changes first and then I’ll fix the bugs today.

**Comment by eXponenta** [comment]
> @eXponenta I already know. Review the types of changes first and then I’ll fix the bugs today.

How should we review it when it doesn't works properly?

**Comment by SukantPal** [comment]
@eXponenta Sometimes people don’t like me changing something even if it works

**Comment by ivanpopelyshev** [comment]
Don't worry, we'll take care of it in weekend together.

**Comment by SukantPal** [comment]
@ivanpopelyshev Any updates?

**Comment by ivanpopelyshev** [comment]
Cant commit, one test is failing because I added extra condition: two consecutive `getLocalBounds` calls shoould not trigger second `calculateBounds`.

**Comment by SukantPal** [comment]
@ivanpopelshev Will look at it in evening

**Comment by ivanpopelyshev** [comment]
Yeah, if you manage to adjust `boundsID` hacks that way tests passes - it will be awesome!

**Comment by cursedcoder** [comment]
Looks good, but tests.

**Comment by ivanpopelyshev** [comment]
My fault - I made extra test that I thought can be passed with this PR, but it turns out something is wrong with boundsID's. I'm waiting for @SukantPal  when he will be free to investigate whether i made a mistake or he did

**Comment by SukantPal** [comment]
@ivanpopelyshev I pushed changes to make the tests to pass. I following breaking changes will occur:

* Display objects that have a change in their natural bounds will _have to declare that by updating `_boundsID`_. They will not get bounds update for free as before (because `updateTransform` will not be called if bounds are already update-to-date in `getLocalBounds`).

---

* `Graphics.clear()` updates `_boundsID`.

**Comment by bigtimebuddy** [comment]
Could you help me understand the breaking change more? If you could show some code of what would change, that’d be helpful.

**Comment by SukantPal** [comment]
@bigtimebuddy I'll explain the "breaking" change a bit rigorously as follows:

Let's say we have a display-object `object`;  it is (and was) expected that if its transform has not changed and `object._boundsID` is not dirty then`object.getBounds()` will calculate the same bounds.

Now, the local bounds are, by definition, defined as bounds _without any transforms applied_. That leaves us with: if `object._boundsID` is not marked dirty, then `object.getLocalBounds()` will calculate the same value.

---

Previously, `getLocalBounds` always deferred to `getBounds` - which always did an `updateTransform`. Now, `updateTransform` marks `_boundsID` dirty, well because it thinks that the transform has changed.

Hence, since `_boundsID` was made dirty internally in `getLocalBounds` - that meant `calculateBounds` was called each and every time `getLocalBounds` was called, and that authors of display-objects (like `PIXI.Graphics#clear()`) did not have to worry about marking their bounds dirty.

**This assumption is now incorrect with this PR**. If `object._boundsID` is in-sync with `this._bounds.updateID`, then the local bounds won't be updated.

---

## Why?

By _not_ updating the local bounds when they aren't marked dirty, we prevent unnecessary bounds calculations and transform updates.

## Future work

Similarly, we could prevent `getBounds()` updates if the transform itself has not changed after a `updateTransform`. That would be done in a different PR.

## Another use-case of this PR

When the scene graph is being rendered, it is expected that `this._bounds` are update-to-date. However, calling `getLocalBounds()` previously changed what `getBounds(true)` returned. This is because `getLocalBounds()` would store its result in `this._bounds` and `getBounds(true)` wouldn't update `_bounds` (no transform update). This results in many errors throughout PixiJS (because the bounds returned by the node are incorrect).

This PR swaps `this._bounds` with another `this._localBounds`. This way `getLocalBounds` will not affect the result of `getBounds(true)`. Sadly though, it is still unable to prevent `getLocalBounds` from affecting the `getBounds(true)` value of grandchildren (and nodes below 1st generation children). I will work on a performant fix for that later on.

**Comment by SukantPal** [comment]
@bigtimebuddy We have a problem with parent containers whose local bounds depend on the transform of children in the parent’a reference frame.

I will push a fix for that, if possible.

**Comment by bigtimebuddy** [comment]
@SukantPal, I think the work that you've done here is terrific. I commend you for optimizing these code paths related to calculating bounds, because it's subtle, difficult work.

One of the challenges of PRs like this, is that I (and probably others on the PixiJS team), don't fully understand the extent to which this may produce unwelcome side-effects for developers. It's one thing add unit-tests, it's another to do a full integration where users heavily rely on bounds.

Our goal as an open source library is, most importantly, is to make sure we do no harm. I think it's important that you can work with developers (@themoonrat, @GoodBoyDigital, and others) to test this build into real-world, complex apps and games. As much as you're able to prove and show that this will not negatively impact users, it will be easier to approve.

**Comment by SukantPal** [comment]
@bigtimebuddy I have removed the sync barrier so there are no side-effects. I think @ivanpopelyshev would be okay to merge this now. (So this PR now only prevents getLocalBounds from storing its result in this._bounds, preventing corruption this.getBounds(true), via the _localBounds swap).

To prevent unnecessary bounds updates, we will need to implement a system where transform updates are propagated up the scene graph. This will be experimented on another branch.

**Comment by ivanpopelyshev** [comment]
There's only one thing left - `_localBounds` is still created in constructor, and its triggered only by calling `getlocalBounds` of the same element - it better be a lazy variable.

**Comment by GoodBoyDigital** [comment]
Hey @SukantPal , thanks for this! Going to test this branch on one of our more layout heavy games! Will let you know how it goes :D

**Pull request merged**